pkg/pointerdb/pointerdb.go

```diff
@@ -131,136 +129,48 @@ func (s *Server) Get(ctx context.Context, req *pb.GetRequest) (resp *pb.GetRespo
-	metaFlags uint32) (items []*pb.ListResponse_Item, more bool) {
-	skip := startAfter != ""
-	startAfterPath := prefix.Append(startAfter)
-	endBeforePath := prefix.Append(endBefore)
-
-	for _, key := range keys {
-		p := paths.New(string(key))
-
-		if skip {
-			if reflect.DeepEqual(p, startAfterPath) {
-				// TODO(kaloyan): Better check - what if there is no path equal to startAfter?
-				// TODO(kaloyan): Add Equal method in Path type
-				skip = false
-			}
-			continue
-		}
-
-		// TODO(kaloyan): Better check - what if there is no path equal to endBefore?
-		// TODO(kaloyan): Add Equal method in Path type
-		if reflect.DeepEqual(p, endBeforePath) {
-			break
-		}
-
-		if !p.HasPrefix(prefix) {
-			// We went through all keys that start with the prefix
-			break
-		}
-
-		if !recursive && len(p) > len(prefix)+1 {
-			continue
-		}
-
-		item := s.createListItem(ctx, p, metaFlags)
-		items = append(items, item)
-
-		if len(items) == limit {
-			more = true
-			break
-		}
+// createListItem creates a new list item with the given path. It also adds
+// the metadata according to the given metaFlags.
+func (s *Server) createListItem(rawItem storage.ListItem, metaFlags uint32) *pb.ListResponse_Item {
+	item := &pb.ListResponse_Item{
+		Path:     rawItem.Key.String(),
+		IsPrefix: rawItem.IsPrefix,
 	}
-	return items, more
-}
-
-// processKeysBackwards iterates backwards through given keys, and returns them
-// as list items
-func (s *Server) processKeysBackwards(ctx context.Context, keys storage.Keys,
-	prefix paths.Path, endBefore string, recursive bool, limit int,
-	metaFlags uint32) (items []*pb.ListResponse_Item, more bool) {
-	skip := endBefore != ""
-	endBeforePath := prefix.Append(endBefore)
-
-	for i := len(keys) - 1; i >= 0; i-- {
-		key := keys[i]
-		p := paths.New(string(key))
-
-		if skip {
-			if reflect.DeepEqual(p, endBeforePath) {
-				// TODO(kaloyan): Better check - what if there is no path equal to endBefore?
-				// TODO(kaloyan): Add Equal method in Path type
-				skip = false
-			}
-			continue
-		}
-
-		if !p.HasPrefix(prefix) || len(p) <= len(prefix) {
-			// We went through all keys that start with the prefix
-			break
-		}
-
-		if !recursive && len(p) > len(prefix)+1 {
-			continue
-		}
-
-		item := s.createListItem(ctx, p, metaFlags)
-		items = append([]*pb.ListResponse_Item{item}, items...)
-
-		if len(items) == limit {
-			more = true
-			break
-		}
+	if item.IsPrefix {
+		return item
 	}
-	return items, more
-}
 
-// createListItem creates a new list item with the given path. It also adds
-// the metadata according to the given metaFlags.
-func (s *Server) createListItem(ctx context.Context, p paths.Path,
-	metaFlags uint32) *pb.ListResponse_Item {
-	item := &pb.ListResponse_Item{Path: p.String()}
-	err := s.getMetadata(ctx, item, metaFlags)
+	err := s.setMetadata(item, rawItem.Value, metaFlags)
 	if err != nil {
 		s.logger.Warn("err retrieving metadata", zap.Error(err))
```

Comment:
should we return the err here?

Reply:
i think if we return something, we don't need to log it.

Reply:
that's my opinion, leave logging for a higher level when possible

Reply:
Here it's a warning however the listing of item can still succeed.